manager/allocator/network/port/port.go

```diff
@@ -0,0 +1,429 @@
+		}
+		// if the published port of the spec is given by the user, we're done
+		if spec.PublishedPort != 0 {
+			continue
+		}
+		// check if there is an old spec. If not, we don't have to go looking
+		// for a previously assigned dynamic port
+		if endpoint.Spec == nil {
+			continue
+		}
+		// go through all of the ports in the old spec, checking if one matches
+		for _, old := range endpoint.Spec.Ports {
+			if portsEqual(old, spec) {
+				// if the do match, find the port from the object
+				for _, p := range endpoint.Ports {
+					if PortsMostlyEqual(spec, p) {
+						// and take it's PublishedPort assignment for the final
+						// port assignment
+						finalPorts[i].PublishedPort = p.PublishedPort
+					}
+				}
+			}
+		}
+	}
+
+	// to avoid altering the "real" state of the ports map, we use the proposal
+	// object, which contains the proposed changes to the Allocator. this
+	// means if there is a failure in the caller after calling Allocate, the
+	// caller can discard the changes
+	prop := &proposal{
+		pa:         pa,
+		allocate:   make(map[port]struct{}, len(finalPorts)),
+		deallocate: make(map[port]struct{}, len(endpoint.Ports)),
+	}
+
+	// now, deallocate everything in the old object
+	for _, p := range endpoint.Ports {
+		// skip publish mode ports
+		if p.PublishMode == api.PublishModeHost {
+			continue
+		}
+
+		prop.deallocate[port{p.Protocol, p.PublishedPort}] = struct{}{}
+	}
+
+	// and then allocate everything in the new object
+
+	// we'll do this in two steps. in the first step, allocate any ports that
+	// have a published port assigned already. this first step prevents us from
+	// choosing a published port for one port that another, later port wants to
+	// use
+	for _, p := range finalPorts {
+		// Skip all Host ports, which we take no action on
+		if p.PublishedPort == 0 || p.PublishMode == api.PublishModeHost {
+			continue
+		}
+		portObj := port{p.Protocol, p.PublishedPort}
+		if _, ok := pa.ports[portObj]; ok {
+			// check if we're deallocating this port
+			if _, ok := prop.deallocate[portObj]; !ok {
+				return nil, ErrPortInUse{portObj}
+			}
+		}
+		// also check that we haven't already tried to allocate this port for
+		// this particular endpoint
+		if _, ok := prop.allocate[portObj]; ok {
+			return nil, ErrPortInUse{portObj}
+		}
+
+		// now, mark this port as "in use" in the newPorts map.
+		prop.allocate[portObj] = struct{}{}
+	}
+
+	// now, this second go around, we'll choose all new publish ports to
+	// dynamically allocate for any port that doesn't have a publish port yet
+ports:
+	for _, p := range finalPorts {
+		// Again, skip all host ports
+		if p.PublishedPort != 0 || p.PublishMode == api.PublishModeHost {
+			continue
+		}
+		portObj := port{p.Protocol, DynamicPortStart}
+		// loop through the whole range of dynamic ports and select the first
+		// one available
+		for i := DynamicPortStart; i <= DynamicPortEnd; i++ {
+			portObj.port = i
+			if _, ok := pa.ports[portObj]; !ok {
```

Comment:
This doesn't seem to take into account any deallocated ports.  If we're near almost exhausting the port space, and this update deallocates 3 ports but needs to allocate 3 different ports, we should be able to do so (at least, by share number) but we can't,  since this loop isn't checking to see if the taken port has been deallocated.

I can also see how maybe that makes sense to do because if we're changing the name of the port or something else about the port config, etc., maybe it'd be confusing if the changed port config gets dynamically allocated the same port on update.

Reply:
Actually, yeah, you're right. I'll need to figure out the correct solution.

Reply:
Fixed this issue by adding another loop further down checking if any dynamic ports are freed.

Reply:
Totally non-blocking nit - that loop looks fine and definitely fixes the issue - I was just wondering if it'd be easier to understand as an extra else block here?  Since this loop already iterates over only the dynamic ports, that would mean that you may not need an extra check for whether the deallocated port is in the dynamic ports?

Also you already have a port object you can use to index into the `deallocate` map.

This is probably just a disfunction of how my brain works, but I find `goto` (or `continue <label>`) control flow a little more confusing to follow than just if else blocks.

Reply:
doing this in two loops preserves the behavior of preferring to not reassign the same dynamic ports that were just freed over again. i don't think that that behavior is _required_, and i don't want to make that behavior _part of the api contract_, but it is rather nice. i can add a comment explaining that we try this but not to rely on it.

> Also you already have a port object you can use to index into the deallocate map.

I'm brain farting on what you mean by this.

You're right about `continue`, though, it is a slightly-less-harmful goto.

Reply:
> doing this in two loops preserves the behavior of preferring to not reassign the same dynamic ports that were just freed over again. i don't think that that behavior is required, and i don't want to make that behavior part of the api contract, but it is rather nice. i can add a comment explaining that we try this but not to rely on it.

Ah that makes sense.

> I'm brain farting on what you mean by this.

Er sorry, I meant if you had the one loop, you could just index into `deallocated` with https://github.com/docker/swarmkit/pull/2579/files#diff-50dd126d07c8b730e5bae674f9542aa2R384.

Not really relevant if we want a separate loop as you say to try to change the dynamically allocated ports each time.